stats/opentelemetry/internal/grpc_trace_bin_propagator_test.go

```diff
@@ -0,0 +1,293 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package internal
+
+import (
+	"context"
+	"encoding/base64"
+	"reflect"
+	"testing"
+
+	otelpropagation "go.opentelemetry.io/otel/propagation"
+	oteltrace "go.opentelemetry.io/otel/trace"
+	"google.golang.org/grpc/internal/grpctest"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/stats"
+	itracing "google.golang.org/grpc/stats/opentelemetry/internal/tracing"
+)
+
+// TODO: Move out of internal as part of open telemetry API
+
+// validSpanContext is a valid OpenTelemetry span context.
+var validSpanContext = oteltrace.SpanContext{}.WithTraceID(
+	oteltrace.TraceID{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16}).WithSpanID(
+	oteltrace.SpanID{17, 18, 19, 20, 21, 22, 23, 24}).WithTraceFlags(
+	oteltrace.TraceFlags(1))
+
+type s struct {
+	grpctest.Tester
+}
+
+func Test(t *testing.T) {
+	grpctest.RunSubTests(t, s{})
+}
+
+// TestInject verifies that the GRPCTraceBinPropagator correctly injects
+// OpenTelemetry span context as `grpc-trace-bin` header in the provided
+// carrier's context, for both fast and slow path, if span context is valid. If
+// span context is invalid, carrier's context is not modified which is verified
+// by retrieving zero value span context from carrier's context.
+//
+// For fast path, it passes `CustomCarrier` as carrier to `Inject()` which
+// injects the span context using `CustomCarrier.SetBinary()`. It then
+// retrieves the injected span context using `stats.OutgoingTrace()` for
+// verification because `SetBinary()` does injection directly in binary format.
+//
+// For slow path, it passes `otel.MapCarrier` as carrier to `Inject()` which
+// injects the span context after base64 encoding as string using
+// `carrier.Set()`. It then retrieves the injected span context using
+// `carrier.Get()` for verification because `Set()` does injection in string
+// format.
+func (s) TestInject(t *testing.T) {
+	tests := []struct {
+		name    string
+		sc      oteltrace.SpanContext
+		fast    bool
+		validSC bool
+	}{
+		{
+			name:    "fast path, valid context",
+			sc:      validSpanContext,
+			fast:    true,
+			validSC: true,
+		},
+		{
+			name:    "fast path, invalid context",
+			sc:      oteltrace.SpanContext{},
+			fast:    true,
+			validSC: false,
+		},
+		{
+			name:    "slow path, valid context",
+			sc:      validSpanContext,
+			fast:    false,
+			validSC: true,
+		},
+		{
+			name:    "slow path, invalid context",
+			sc:      oteltrace.SpanContext{},
+			fast:    false,
+			validSC: false,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			p := GRPCTraceBinPropagator{}
+			tCtx, tCancel := context.WithCancel(context.Background())
+			tCtx = oteltrace.ContextWithSpanContext(tCtx, test.sc)
+			defer tCancel()
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
```

Comment:
Why have two "root" contexts instead of just deriving the otel context from ctx?

Reply:
So, `tCtx` is the trace context that is passed to propagator to inject and `ctx` is the regular context with which carrier is initialized. When we call `propagator.Inject(tCtx, carrier)`, span context is extracted from `tCtx` and injected into carrier's context which is what we have to verify.

Reply:
Discussed offline. Moved under opentelemetry package where it has to be eventually